apps/api/src/app/events/usecases/message-matcher/message-matcher.usecase.ts

```diff
@@ -225,6 +238,115 @@ export class MessageMatcher {
     ));
   }
 
+  private async processHaveSeen(
+    filter: IPreviousStepFilterPart,
+    command: SendMessageCommand,
+    filterProcessingDetails: FilterProcessingDetails
+  ) {
+    const previousJobId = command.job._parentId;
+
+    if (!previousJobId) {
+      return true;
+    }
+    const message = await this.messageRepository.findOne({
+      _jobId: previousJobId,
+      _organizationId: command.organizationId,
+      _environmentId: command.environmentId,
+    });
+
+    if (!message) {
+      return true;
+    }
+    if (message?.channel === ChannelTypeEnum.IN_APP) {
+      filterProcessingDetails.addCondition({
+        filter: FILTER_TO_LABEL[filter.on],
+        field: 'read',
+        expected: 'true',
+        actual: `${message.seen}`,
+        operator: 'EQUAL',
+        passed: message.seen,
+      });
+
+      return message.seen;
+    }
+
+    const count = await this.executionDetailsRepository.count({
+      _jobId: previousJobId,
+      _messageId: message._id,
+      _environmentId: command.environmentId,
+      _organizationId: command.organizationId,
+      webhookStatus: {
+        $in: [EmailEventStatusEnum.OPENED],
+      },
+    });
+
+    filterProcessingDetails.addCondition({
+      filter: FILTER_TO_LABEL[filter.on],
+      field: 'read',
+      expected: 'true',
+      actual: `${count > 0}`,
+      operator: 'EQUAL',
+      passed: count > 0,
+    });
+
+    return count > 0;
+  }
+
+  private async processHaveRead(
+    filter: IPreviousStepFilterPart,
+    command: SendMessageCommand,
+    filterProcessingDetails: FilterProcessingDetails
+  ) {
+    const previousJobId = command.job._parentId;
+
+    if (!previousJobId) {
+      return true;
+    }
+
+    const message = await this.messageRepository.findOne({
+      _jobId: previousJobId,
+      _organizationId: command.organizationId,
+      _environmentId: command.environmentId,
+    });
+
+    if (!message) {
+      return true;
+    }
+    if (message?.channel === ChannelTypeEnum.IN_APP) {
+      filterProcessingDetails.addCondition({
+        filter: FILTER_TO_LABEL[filter.on],
+        field: 'read',
+        expected: 'true',
+        actual: `${message.read}`,
+        operator: 'EQUAL',
+        passed: message.read,
+      });
+
+      return message.read;
+    }
+
+    const count = await this.executionDetailsRepository.count({
+      _jobId: previousJobId,
+      _messageId: message._id,
+      _environmentId: command.environmentId,
+      _organizationId: command.organizationId,
+      webhookStatus: {
+        $in: [EmailEventStatusEnum.OPENED],
```

Comment:
There are few (I think 3) differences between `processHaveRead` and `processHaveSeen`. Line 320, 322, 334.
I would totally have a shared function to reuse the code. But I can get it might hard. Though the pattern we have changing `filterProcessingDetails` inside of the functions is what makes more complicated to have an easy reusable logic.